pandas/io/pytables.py

```diff
@@ -1778,10 +1780,11 @@ def __ne__(self, other) -> bool:
     @property
     def is_indexed(self) -> bool:
         """ return whether I am an indexed column """
-        try:
-            return getattr(self.table.cols, self.cname).is_indexed
-        except AttributeError:
+        if not hasattr(self.table, "cols"):
+            # e.g. if self.set_table hasn't been called yet, self.table
+            #  will be None.
             return False
+        return getattr(self.table.cols, self.cname).is_indexed  # type: ignore
```

Comment:
@simonjayhawkins I expected the hasattr check above to be enough to make this `# type: ignore` unnecessary, but it wasn't.  any ideas?

Reply:
see https://github.com/python/mypy/issues/1424

best to add the mypy error message as a comment and also a link to the mypy issue.

_might_ also be best to split the return statement to isolate the ignore to the result of the getattr expression.

then casting as suggested in the mypy issue would still have the return type of is_indexed of the casted type checked against the return type of this method. 
This could result in less clear code so will leave it to you to decide whether you think it warrants changing the code here.

This could also be done my error codes, see #29197, but needs more discussion

Reply:
comment added

Reply:
sorry I couldn't look in detail yesterday and just gave a generic answer.

the error message is `error: "None" has no attribute "cols"`.  

`reveal_type(self.table)` gives `note: Revealed type is 'None'`

in `IndexCol.__init__`, `self.table = None`

I _think_ if you add the type annotation for `self.table` and add `assert self.table is not None` before the return statement, we can avoid the ignore.

(in general if you can include the exact mypy error message as a comment makes it easier/quicker to review. We do this in other places)

Reply:
> I think if you add the type annotation for self.table and add assert self.table is not None before the return statement, we can avoid the ignore.

AFAICT this still doesnt work because the Table class still doesnt have a "cols" attr.

> (in general if you can include the exact mypy error message as a comment makes it easier/quicker to review. We do this in other places)

Will update.  Thanks for being patient with me

Reply:
> AFAICT this still doesnt work because the Table class still doesnt have a "cols" attr.

fair enough. the generic comment above is valid. a type ignore is likely required in many places where we are using hasattr.

I thought that maybe the type of self.table would be a type with a "cols" attr. and that the hasattr was used to allow duck typing of other classes with a "cols" attr.

If this were the case mypy would only be checking the nominal type defined for self.table ( and None)